sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala

```diff
@@ -589,22 +589,43 @@ case class DateFormatClass(left: Expression, right: Expression, timeZoneId: Opti
 
   override def inputTypes: Seq[AbstractDataType] = Seq(TimestampType, StringType)
 
-  override def withTimeZone(timeZoneId: String): TimeZoneAwareExpression =
+  var formatter: Option[TimestampFormatter] = None
+
+  override def withTimeZone(timeZoneId: String): TimeZoneAwareExpression = {
+    if (formatter.isEmpty && right.foldable) {
+      val format = right.eval().toString
+      formatter = Some(TimestampFormatter(
+        format,
+        DateTimeUtils.getZoneId(timeZoneId),
+        Locale.US))
+    }
     copy(timeZoneId = Option(timeZoneId))
+  }
 
   override protected def nullSafeEval(timestamp: Any, format: Any): Any = {
-    val df = TimestampFormatter(format.toString, zoneId)
-    UTF8String.fromString(df.format(timestamp.asInstanceOf[Long]))
+    val tf = if (formatter.isEmpty) {
```

Comment:
How about `.getOrElse`?

Reply:
`.getOrElse` has some overhead of calling the lambda function. I explicitly avoided its usage in the interpreted mode. For consistency, I could do the same in the codegen function but I don't think it does matter.